File: programs/format_tutorials.py
#!/usr/bin/env python3
"""
Auto-formats our tutorials, saving manual formatting work:

- Converts space-based indentations to tabs in code blocks.
- Fills GDScript code comments as paragraphs.
- Wraps symbols and numeric values in code.
- Wraps other capitalized names, pascal case values into italics (we assume they're node names).
- Marks code blocks without a language as using `gdscript`.
- Add <kbd> tags around keyboard shortcuts (the form needs to be Ctrl+F1).
"""
import argparse
import itertools
import logging
import os
import re
import sys
import textwrap
from dataclasses import dataclass
from typing import List

import colorama
from lib.gdscript_classes import BUILT_IN_CLASSES

TAB_WIDTH: int = 4

LOGGER = logging.getLogger("format_tutorial.py")

ERROR_PYTHON_VERSION_TOO_OLD: int = 1
ERROR_INCORRECT_FILE_PATHS: int = 2

WORDS_TO_KEEP_UNFORMATTED: List[str] = [
    "gdquest",
    "gdscript",
    "godot",
    "stack overflow",
    "google",
    "youtube",
    "twitter",
    "facebook",
    "discord",
    "instagram",
    "duckduckgo",
]

RE_SPLIT_CODE_BLOCK: re.Pattern = re.compile(r"(```[a-z]*.*?```)", flags=re.DOTALL)
RE_SPLIT_HTML: re.Pattern = re.compile(r"(<.+?>.*?<\/.+?>)", flags=re.DOTALL)
RE_SPLIT_TEMPLATE: re.Pattern = re.compile(r"({%.+?%})")
RE_SPLIT_FRONT_MATTER: re.Pattern = re.compile(r"(---.+?---\n)", flags=re.DOTALL)
RE_BUILT_IN_CLASSES: re.Pattern = re.compile(
    r"\b(?<!`)({})\b".format(r"|".join(BUILT_IN_CLASSES))
)
# Matches paths with a filename at the end.
PATTERN_DIR_PATH: str = r"\b((res|user)://)?/?(([\w]+/)+(\w*\.\w+)?\b)"
PATTERN_FILE_AT_ROOT: str = r"(res|user)://\w+\.\w+"
SUPPORTED_EXTENSIONS: List[str] = [
    "png",
    "jpe?g",
    "mp4",
    "mkv",
    "t?res",
    "t?scn",
    "gd",
    "py",
    "shader",
]
PATTERN_FILENAME_ONLY: str = r"(\w+\.({}))".format("|".join(SUPPORTED_EXTENSIONS))
RE_FILE_PATH: re.Pattern = re.compile(
    "|".join([PATTERN_DIR_PATH, PATTERN_FILE_AT_ROOT, PATTERN_FILENAME_ONLY])
)
# Matches directory paths without a filename at the end. Group 1 targets the path.
#
# Known limitations:
# - The path requires a trailing slash followed by a space, or period and space,
# or the line ends with a period.
# - Won't capture a leading slash.
SINGLE_FUNCTION_CALL: str = r"\b_?\w+(\.\w+)*\([\"'_a-zA-Z0-9, ]*\)"
SINGLE_VARIABLE: str = r"\b_\w+|\b[a-zA-Z0-9]+_\w+"
VARIABLE_AND_PROPERTY = r"\b_?\w+\.\w+"
RE_VARIABLE_OR_FUNCTION: re.Pattern = re.compile(
    "|".join([SINGLE_FUNCTION_CALL, SINGLE_VARIABLE, VARIABLE_AND_PROPERTY])
)
RE_NUMERIC_VALUES_AND_RANGES: re.Pattern = re.compile(
    r"(\d+x\d+)|(\[[\d\., ]+\])|(-?\d+\.\d+)|(?<![\nA-Za-z])(-?\d+)(?![A-Za-z])(?!\. )"
)
# Sequence of multiple words with an optional "->" separator, to italicize.
RE_TO_ITALICIZE_SEQUENCE: re.Pattern = re.compile(
    r"(?<![\-\.] )(?<!^)[A-Z0-9]+[a-zA-Z0-9]*( (-> )?[A-Z][a-zA-Z0-9]+(\.\.\.)?)+",
    flags=re.MULTILINE,
)
# Capitalized words and PascalCase that are not at the start of a sentence or a line.
RE_TO_ITALICIZE_ONE_WORD: re.Pattern = re.compile(
    r"^([A-Z]\w+[A-Z]\w+)|(?<![\.\-:;?!] )(?<!^)([A-Z][a-zA-Z0-9]+(\.\.\.)?)",
    flags=re.MULTILINE,
)
RE_TO_IGNORE: re.Pattern = re.compile(r"(!?\[.*\]\(.+\)|^#+ .+$)", flags=re.MULTILINE)
RE_KEYBOARD_SHORTCUTS: re.Pattern = re.compile(
    r"(?<!\d\. ) +((Ctrl|Alt|Shift|CTRL|ALT|SHIFT) ?\+ ?)(F\d{1,2}|[A-Z0-9])"
)
RE_KEYBOARD_SHORTCUTS_ONE_ELEMENT: re.Pattern = re.compile(
    r"Ctrl|Alt|Shift|CTRL|ALT|SHIFT|[A-Z0-9]+"
)
RE_HEX_VALUES: re.Pattern = re.compile(r"#[a-fA-F0-9]{3,8}")
RE_INSIDE_DOUBLE_QUOTES: re.Pattern = re.compile(r'[^\w]("[^"]*")[^\w]')
RE_INSIDE_SINGLE_QUOTES: re.Pattern = re.compile(r"[^\w]('[^']*')[^\w]")
RE_MARKDOWN_BLOCKQUOTE: re.Pattern = re.compile(r"^(> )(.+?)$", flags=re.MULTILINE)


@dataclass
class ProcessedDocument:
    """Maps a file path to formatted content"""

    file_path: str
    content: str


def inline_code_built_in_classes(text: str) -> str:
    return RE_BUILT_IN_CLASSES.sub(lambda match: "`{}`".format(match.group(0)), text)


def inline_code_paths(text: str) -> str:
    return RE_FILE_PATH.sub(lambda match: "`{}`".format(match.group(0)), text)


def inline_code_variables_and_functions(text: str) -> str:
    return RE_VARIABLE_OR_FUNCTION.sub(
        lambda match: "`{}`".format(match.group(0)), text
    )


def inline_code_hex_values(text: str) -> str:
    return RE_HEX_VALUES.sub(lambda match: "`{}`".format(match.group(0)), text)


def inline_code_numeric_values(text: str) -> str:
    return RE_NUMERIC_VALUES_AND_RANGES.sub(
        lambda match: "`{}`".format(match.group(0)),
        text,
    )


def replace_double_inline_code_marks(text: str) -> str:
    """Finds and replaces cases where we have `` to `."""
    return re.sub("(`+\b)|(\b`+)", "`", text)


def italicize_word_sequences(text: str) -> str:
    return RE_TO_ITALICIZE_SEQUENCE.sub(
        lambda match: "*{}*".format(match.group(0)), text
    )


def italicize_other_words(text: str) -> str:
    def replace_match(match: re.Match) -> str:
        expression: str = match.group(0)
        if (
            expression.lower() in WORDS_TO_KEEP_UNFORMATTED
            or expression.upper() == expression
        ):
            return expression
        return "*{}*".format(match.group(0))

    return RE_TO_ITALICIZE_ONE_WORD.sub(replace_match, text)


def add_keyboard_tags(text: str) -> str:
    def add_one_keyboard_tag(match: re.Match) -> str:
        expression = match.group(0)
        if expression.strip() == "I":
            return expression
        return RE_KEYBOARD_SHORTCUTS_ONE_ELEMENT.sub(
            lambda m: "<kbd>{}</kbd>".format(m.group(0)),
            expression,
        )

    return RE_KEYBOARD_SHORTCUTS.sub(add_one_keyboard_tag, text)


FORMATTERS = [
    inline_code_paths,
    inline_code_variables_and_functions,
    italicize_word_sequences,
    add_keyboard_tags,
    inline_code_built_in_classes,
    inline_code_hex_values,
    italicize_other_words,
    inline_code_numeric_values,
]


def format_line(line: str) -> str:
    def split_formatted_words(text: str) -> List[str]:
        return [s for s in re.split(r"([`*].+?[`*])", text) if s != ""]

    out: str = line
    split_line: List[str] = split_formatted_words(line)

    if len(split_line) > 1:
        out = "".join(list(map(format_line, split_line)))

    elif not re.match("^[`*]", line):
        for formatter in FORMATTERS:
            formatted_line: str = formatter(line)
            if formatted_line == line:
                continue

            split_line: List[str] = split_formatted_words(formatted_line)
            if len(split_line) > 1:
                out = "".join(list(map(format_line, split_line)))
            else:
                out = formatted_line
            break
    return out


def format_content(content: str) -> str:
    """Applies styling rules to content other than a code block."""
    return "\n".join(list(map(format_line, content.split("\n"))))


def format_code_block(text: str):
    """Applies styling rules to one code block"""

    def convert_spaces_to_tabs(content: str) -> str:
        return content.replace(" " * TAB_WIDTH, "\t")

    def fill_comment(match: re.Match, line_length: int = 80) -> str:
        """Takes one line of comment and wraps it at the `line_length` column."""

        def count_indents(text: str) -> int:
            count = 0
            while text[count] == "\t":
                count += 1
            return count

        text = match.group(0)
        indent_level = count_indents(text)
        # We need to pad every line with that many indents, comment signs, and one space.
        # So we take it into account before wrapping
        dash_count = text.count("#", indent_level, indent_level + 2)
        prefix_width = TAB_WIDTH * indent_level + dash_count + 1
        wrap_length = line_length - prefix_width

        trimmed_text = text.lstrip("\t# ")

        wrapped_text = textwrap.wrap(trimmed_text, wrap_length)
        output = [
            "\t" * indent_level + "#" * dash_count + " " + line for line in wrapped_text
        ]
        return "\n".join(output)

    match = re.match("```([a-z]+)?\n(.*?)```", text, flags=re.DOTALL)

    language = match.group(1) or "gdscript"

    content = match.group(2)
    content = convert_spaces_to_tabs(content)
    content = re.sub("\t*#.+", fill_comment, content, flags=re.MULTILINE)

    output = "```{}\n{}```".format(language, content)
    return output


def parse_command_line_arguments(args) -> argparse.Namespace:
    """Parses the command line arguments"""
    parser = argparse.ArgumentParser(
        description=__doc__,
    )
    parser.add_argument(
        "files",
        type=str,
        nargs="+",
        default="",
        help="A list of paths to markdown files.",
    )
    parser.add_argument(
        "-o",
        "--output",
        type=str,
        default="",
        help="Path to the output directory.",
    )
    parser.add_argument(
        "-i", "--in-place", action="store_true", help="Overwrite the source files."
    )
    return parser.parse_args(args)


def process_content(content: str) -> str:
    """Applies formatting rule to a file's content."""

    def flatten(items):
        for item in items:
            if isinstance(item, list) and not isinstance(item, (str, bytes)):
                yield from flatten(item)
            else:
                yield item

    def is_code_block(text: str) -> bool:
        return text.startswith("```")

    def is_unformatted_block(text: str) -> bool:
        """Returns true if the text starts like HTML, a template element, a
        blockquote, or a quote symbol."""
        ignore_pairs = [
            ("```", "```"),
            ("<", ">"),
            ("---", "---"),
            ("{%", "%}"),
            ("'", "'"),
            ('"', '"'),
        ]
        for opening, closing in ignore_pairs:
            if text.lstrip().startswith(opening) and text.rstrip().endswith(closing):
                return True
        return False

    def split_by_regex(text: str, regex: str) -> List[str]:
        if is_unformatted_block(text):
            return [text]
        return re.split(regex, text)

    sections = RE_SPLIT_CODE_BLOCK.split(content)
    for regex in [
        RE_SPLIT_FRONT_MATTER,
        RE_SPLIT_HTML,
        RE_SPLIT_TEMPLATE,
        RE_INSIDE_DOUBLE_QUOTES,
        RE_INSIDE_SINGLE_QUOTES,
        RE_MARKDOWN_BLOCKQUOTE,
    ]:
        sections = map(split_by_regex, sections, itertools.repeat(regex))
        sections = list(flatten(sections))
    sections = list(filter(lambda text: bool(text), sections))

    formatted_sections: List[str] = []

    for block in sections:
        if is_code_block(block):
            formatted_sections.append(format_code_block(block))
        elif is_unformatted_block(block):
            formatted_sections.append(block)
        else:
            chunks: List[str] = re.split(RE_TO_IGNORE, block)
            formatted_chunks: List[str] = []
            for chunk in chunks:
                if re.match(RE_TO_IGNORE, chunk):
                    formatted_chunks.append(chunk)
                else:
                    formatted_chunks.append(format_content(chunk))
            formatted_sections.append("".join(formatted_chunks))

    return "".join(formatted_sections)


def output_result(args: argparse.Namespace, document: ProcessedDocument) -> None:
    """Outputs the content of each processed document either to:

    - The input file if the program was called with the `--in-place` option.
    - A new file if the program was called with the `--output` option.
    - Otherwise, to the standard output.

    """
    if args.in_place:
        with open(document.file_path, "w") as output_file:
            output_file.write(document.content)
    elif args.output == "":
        print(document.content)
    else:
        output_path = os.path.join(args.output, os.path.basename(document.file_path))
        if not os.path.isdir(args.output):
            os.makedirs(args.output)
        with open(output_path, "w") as output_file:
            output_file.write(document.content)


def print_error(*args, **kwargs):
    print(colorama.Fore.RED, end="", flush=True)
    print(*args, file=sys.stderr, **kwargs)
    print(colorama.Fore.RESET, end="", flush=True)


def main():
    def is_python_version_compatible() -> bool:
        return sys.version_info.major == 3 and sys.version_info.minor >= 8

    args: argparse.Namespace = parse_command_line_arguments(sys.argv[1:])
    logging.basicConfig(level=logging.ERROR)
    if not is_python_version_compatible():
        print_error(
            "\n".join(
                [
                    "Your Python version ({}.{}.{}) is too old.",
                    "Minimum required version: 3.8.",
                    "Please install a more recent Python version.",
                    "Aborting operation.",
                ]
            ).format(
                sys.version_info.major, sys.version_info.minor, sys.version_info.micro
            )
        )
        sys.exit(ERROR_PYTHON_VERSION_TOO_OLD)

    filepaths: List[str] = [
        f for f in args.files if f.lower().endswith(".md") and os.path.exists(f)
    ]
    if len(filepaths) != len(args.files):
        print_error(
            "\n".join(
                [
                    "Some files are missing or their path is incorrect.",
                    "Please ensure there's no typo in the path.",
                    "Aborting operation.",
                ]
            )
        )
        sys.exit(ERROR_INCORRECT_FILE_PATHS)

    documents: List[ProcessedDocument] = []
    for file_path in filepaths:
        with open(file_path, "r") as markdown_file:
            content: str = markdown_file.read()
            documents.append(ProcessedDocument(file_path, process_content(content)))
    list(map(output_result, itertools.repeat(args), documents))


if __name__ == "__main__":
    main()

File: format_tutorials/format_tutorials.nim
## Auto-formats GDQuest tutorials, saving manual formatting work:
## 
## - Converts space-based indentations to tabs in code blocks.
## - Fills GDScript code comments as paragraphs.
## - Wraps symbols and numeric values in code.
## - Wraps other capitalized names, pascal case values into italics (we assume they're node names).
## - Marks code blocks without a language as using `gdscript`.
## - Add <kbd> tags around keyboard shortcuts (the form needs to be Ctrl+F1).

import std/parseopt
import std/re
import std/os
import std/sequtils
import std/strutils
import std/wordwrap

import godot_built_ins

const HelpMessage = """Auto-formats markdown documents, saving manual formatting work:

- Converts space-based indentations to tabs in code blocks.
- Fills GDScript code comments as paragraphs.
- Wraps symbols and numeric values in code.
- Wraps other capitalized names, pascal case values into italics (we assume they're node names).
- Marks code blocks without a language as using `gdscript`.
- Add <kbd> tags around keyboard shortcuts (the form needs to be Ctrl+F1).

How to use:

format_tutorials [options] <input-file> [<input-file> ...]

Options:

-i/--in-place: Overwrite the input files with the formatted output.
-o/--output-directory: Write the formatted output to the specified directory.

-h/--help: Prints this help message.
"""


type
    CommandLineArgs = object
        inputFiles: seq[string]
        inPlace: bool
        outputDirectory: string

    BlockKind = enum
        TextParagraph,
        CodeBlock,
        YamlFrontMatter,
        Table,
        List,
        Blockquote,
        Heading,
        EmptyLine,
        Reference,
        Html,
        GDQuestShortcode

    Block = object
        kind: BlockKind
        text: seq[string]


const
    SupportedExtensions = ["png", "jpe?g", "mp4", "mkv", "t?res", "t?scn", "gd", "py", "shader", ]
    PatternFilenameOnly = r"(\w+\.(" & SupportedExtensions.join("|") & "))"
    PatternDirPath = r"((res|user)://)?/?([\w]+/)+(\w*\.\w+)?"
    PatternFileAtRoot = r"(res|user)://\w+\.\w+"
    PatternModifierKeys = r"Ctrl|Alt|Shift|CTRL|ALT|SHIFT|Super|SUPER"
    PatternKeyboardSingleCharacterKey = r"[A-Z0-9!@#$%^&*()_\-{}|\[\]\\;':,\./<>?]"
    PatternOtherKeys = r"F\d{1,2}|Tab|Up|Down|Left|Right|LMB|MMB|RMB|Backspace|Delete"
    PatternFunctionOrConstructorCall = r"\w+(\.\w+)*\(.*?\)"
    PatternVariablesAndProperties = r"_\w+|[a-zA-Z0-9]+([\._]\w+)+"
    PatternGodotBuiltIns = GodotBuiltInClassesByLength.join("|")

let
    RegexFilePath = re([PatternDirPath, PatternFileAtRoot, PatternFilenameOnly].join("|"))
    RegexStartOfList = re"- |\d+\. "
    RegexCodeCommentLine = re"\s*#"
    RegexOnePascalCaseWord = re"[A-Z0-9]\w+[A-Z]\w+|[A-Z][a-zA-Z0-9]+(\.\.\.)?"
    RegexOnePascalCaseWordStrict = re"[A-Z0-9]\w+[A-Z]\w+"
    RegexMenuOrPropertyEntry = re"[A-Z0-9]+[a-zA-Z0-9]*( (-> )+[A-Z][a-zA-Z0-9]+( [A-Z][a-zA-Z0-9]*)*(\.\.\.)?)+"
    RegexCapitalWordSequence = re"[A-Z0-9]+[a-zA-Z0-9]*( (-> )?[A-Z][a-zA-Z0-9]+(\.\.\.)?)+"
    RegexKeyboardShortcut = re("((" & PatternModifierKeys & r") ?\+ ?)+(" & PatternOtherKeys & "|" & PatternKeyboardSingleCharacterKey & ")")
    RegexOneKeyboardKey = re("(" & [PatternModifierKeys, PatternOtherKeys, PatternKeyboardSingleCharacterKey].join("|") & ")")
    RegexNumber = re"\d+(D|px)|\d+(x\d+)*"
    RegexHexValue = re"(0x|#)[0-9a-fA-F]+"
    RegexCodeIdentifier = re([PatternFunctionOrConstructorCall, PatternVariablesAndProperties].join("|"))
    RegexGodotBuiltIns = re(PatternGodotBuiltIns)
    RegexSkip = re"""(_.+?_|\*\*[^*]+?\*\*|\*[^*]+?\*|`.+?`|".+?"|'.+?'|\!?\[.+?\)|\[.+?\])\s*|\s+|$"""
    RegexStartOfSentence = re"\s*\p{Lu}"
    RegexEndOfSentence = re"[.!?:]\s+"


func regexWrap*(regexes: seq[Regex], pair: (string, string)): auto =
    ## Retruns a function that takes the string `text` and finds the first
    ## regex match from `regexes` only at the beginning of `text`.
    ##
    ## The function returns a tuple:
    ## - `(text wrapped by pair, rest of text)` if there's a regex match.
    ## - `("", text)` if there's no regex match.
    return func(text: string): (string, string) =
        for regex in regexes:
            let bound = text.matchLen(regex)
            if bound == -1: continue
            return (pair[0] & text[0 ..< bound] & pair[1], text[bound .. ^1])
        return ("", text)


func regexWrapEach*(regexAll, regexOne: Regex; pair: (string, string)): auto =
    ## Returns a function that takes the string `text` and tries to match
    ## `regexAll` only at the beginning of `text`.
    ##
    ## The function regurns a tuple:
    ## - `(text with all occurances of RegexOne wrapped by pair, rest of text)`
    ##   if there's a regex match.
    ## - `("", text)` if there's no regex match.
    return func(text: string): (string, string) =
        let bound = text.matchLen(regexAll)
        if bound != -1:
            let replaced = replacef(text[0 ..< bound], regexOne, pair[0] & "$1" & pair[1])
            return (replaced, text[bound .. ^1])
        return ("", text)


proc advanceMarkdownTextLineFormatter*(line: string): (string, string) =
    ## Find the first `RegexSkip` and split `line` into the tuple
    ## `(line up to RegexSkip, rest of line)`.
    let (_, last) = line.findBounds(RegexSkip)
    return (line[0 .. last], line[last + 1 .. ^1])


proc formatMarkdownTextLine*(line: string): string =
    ## Returns the formatted `line` using the GDQuest standard.
    let formatters =
        { "any": regexWrapEach(RegexKeyboardShortcut, RegexOneKeyboardKey, ("<kbd>", "</kbd>"))
        , "any": regexWrap(@[RegexFilePath], ("`", "`"))
        , "any": regexWrap(@[RegexMenuOrPropertyEntry], ("*", "*"))
        , "any": regexWrap(@[RegexCodeIdentifier, RegexGodotBuiltIns], ("`", "`"))
        , "any": regexWrap(@[RegexNumber, RegexHexValue], ("`", "`"))
        , "any": regexWrap(@[RegexOnePascalCaseWordStrict], ("*", "*"))
        , "skipStartOfSentence": regexWrap(@[RegexCapitalWordSequence, RegexOnePascalCaseWord], ("*", "*"))
        }
    block outer:
        var
            line = line
            isStartOfSentence = line.startsWith(RegexStartOfSentence)
        while true:
            for (applyAt, formatter) in formatters:
                if line.len <= 0: break outer
                if (applyAt, isStartOfSentence) == ("skipStartOfSentence", true): continue
                let (formatted, rest) = formatter(line)
                result &= formatted
                line = rest

            let (advanced, rest) = advanceMarkdownTextLineFormatter(line)
            isStartOfSentence = advanced.endsWith(RegexEndOfSentence)
            result &= advanced
            line = rest


proc formatMarkdownTextLines*(lines: openArray[string]): string =
    ## Returns the formatted sequence of `lines` using the GDQuest standard.
    var partialResult: seq[string] = @[]
    for line in lines:
        partialResult.add(line.formatMarkdownTextLine)
    return partialResult.join("\n")


proc formatMarkdownList*(lines: seq[string]): string =
    ## Returns the formatted sequence of `lines` using the GDQuest standard.
    ##
    ## `lines` is a sequence representing a markdown list. One item can span
    ## multiple lines, in which case they get concatenated before formatting.
    var
        lines = lines
        linesStart: seq[string] = @[]
        i = 0
    while i < lines.len:
        var bound = lines[i].matchLen(RegexStartOfList)
        if bound != -1:
            linesStart.add(lines[i][0 ..< bound])
            lines[i] = lines[i][bound .. ^1]
            i.inc
        else:
            lines[i - 1] &= " " & lines[i].strip()
            lines.delete(i)

    var partialResult: seq[string] = @[]
    for (lineStart, line) in zip(linesStart, lines.formatMarkdownTextLines.split("\n")):
        partialResult.add(lineStart & line)
    return partialResult.join("\n")


proc formatCodeBlock*(lines: openArray[string]): string =
    ## Returns the formatted `lines` markdown code block using the GDQuest standard:
    ##
    ## - Converts space-based indentations to tabs.
    ## - Fills GDScript code comments as paragraphs with a max line length of 80.
    ## - If the code block has no language set, sets it to "gdscript".
    const TAB_WIDTH = 4
    var formattedStrings: seq[string]

    formattedStrings.add(if lines[0].strip() == "```": "```gdscript" else: lines[0])
    for line in lines[1 .. ^1]:
        var processedLine = line.replace("    ", "\t")
        if not line.match(RegexCodeCommentLine):
            formattedStrings.add(processedLine)
            continue

        var indentCount = 0
        while line[indentCount] == '\t':
            indentCount += 1

        let hashCount = processedLine[indentCount .. indentCount + 2].count("#")
        let margin = indentCount + hashCount + 1
        let desiredTextLength = 80 - indentCount * TAB_WIDTH - hashCount
        let content = wrapWords(line[margin .. ^1], desiredTextLength)
        for line in splitLines(content):
            let optionalSpace = if not line.startswith(" "): " " else: ""
            formattedStrings.add(
                repeat("\t", indentCount) &
                repeat("#", hashCount) &
                optionalSpace &
                line.strip()
             )
    return formattedStrings.join("\n")


proc convertBlocksToFormattedMarkdown*(blocks: seq[Block]): string =
    ## Takes a sequence of `blocks` from a parsed markdown file
    ## and returns a formatted document.
    const IgnoredBlockKinds = [
        EmptyLine, YamlFrontMatter, Blockquote, Heading, Table, GDQuestShortcode, Reference, Html
    ]
    var formattedStrings: seq[string]
    for mdBlock in blocks:
        case mdBlock.kind
        of IgnoredBlockKinds:
            formattedStrings.add(mdBlock.text)
        of CodeBlock:
            formattedStrings.add(formatCodeBlock(mdBlock.text))
        of List:
            formattedStrings.add(formatMarkdownList(mdBlock.text))
        else:
            formattedStrings.add(formatMarkdownTextLines([mdBlock.text.join(" ")]))
    result = formattedStrings.join("\n")
    result.stripLineEnd()


proc parseBlocks*(content: string): seq[Block] =
    ## Parses the markdown document `content` and returns a sequence of blocks.
    ##
    ## This is a limited block parser that ignores many markdown features.
    ## For example, it doesn't parse recursive blocks.
    ##
    ## It's designed to parse the content of GDQuest tutorials for formatting.
    proc createDefaultBlock(): Block =
        return Block(
            kind: TextParagraph,
            text: @[],
        )

    var currentBlock = createDefaultBlock()
    var previousLine = ""
    for line in splitLines(content):
        let
            isEmptyLine = line.strip() == ""
            isClosingParagraph = currentBlock.kind == TextParagraph and currentBlock.text.len() != 0

        if not isEmptyLine:
            currentBlock.text.add(line)

        if isEmptyLine and isClosingParagraph:
            result.add(currentBlock)
            currentBlock = createDefaultBlock()
        # Lines starting with a hash could be comments inside a code block
        elif line.startswith("#") and currentBlock.kind != CodeBlock:
            currentBlock.kind = Heading
            result.add(currentBlock)
            currentBlock = createDefaultBlock()
        elif line.startswith("```"):
            if currentBlock.kind == TextParagraph:
                currentBlock.kind = CodeBlock
            elif currentBlock.kind == CodeBlock:
                result.add(currentBlock)
                currentBlock = createDefaultBlock()
        elif line.startswith("---"):
            if currentBlock.kind == TextParagraph:
                currentBlock.kind = YamlFrontMatter
            elif currentBlock.kind == YamlFrontMatter:
                result.add(currentBlock)
                currentBlock = createDefaultBlock()
        elif line.startsWith("{%") and currentBlock.kind == TextParagraph:
            currentBlock.kind = GDQuestShortcode
        elif line.startsWith("["):
            currentBlock.kind = Reference
            currentBlock = createDefaultBlock()
        elif line.startsWith("<") and currentBlock.kind != Html:
            currentBlock.kind = Html
        elif line.startsWith(">") and currentBlock.kind != Blockquote:
            currentBlock.kind = Blockquote
        elif line.match(RegexStartOfList):
            currentBlock.kind = List
        elif line.startsWith("|"):
            currentBlock.kind = Table

        if isEmptyLine:
            let isClosingBlock = currentBlock.kind in @[List, Html, Table, Blockquote] or
                    (currentBlock.kind == GDQuestShortcode and previousLine.strip().endsWith("%}"))
            if isClosingBlock:
                result.add(currentBlock)
                currentBlock = createDefaultBlock()

            currentBlock.kind = EmptyLine
            currentBlock.text = @[""]
            result.add(currentBlock)
            currentBlock = createDefaultBlock()

        previousLine = line
    # When the file ends, we need to add the last block.
    result.add(currentBlock)


proc formatContent*(content: string): string =
    ## Takes the markdown `content` and returns a formatted document using
    ## the GDQuest standard.
    let blocks = parseBlocks(content)
    return convertBlocksToFormattedMarkdown(blocks)


proc parseCommandLineArguments(): CommandLineArgs =
    for kind, key, value in getopt():
        case kind
        of cmdEnd: break
        of cmdArgument:
            if fileExists(key):
                result.inputFiles.add(key)
            else:
                echo "Invalid filename: ", key
                quit(1)
        of cmdLongOption, cmdShortOption:
            case key
            of "help", "h": echo HelpMessage
            of "in-place", "i": result.inPlace = true
            of "output-directory", "o":
                if isValidFilename(value):
                    result.outputDirectory = value
                else:
                    echo "Invalid output directory: ", value, "\n"
                    echo HelpMessage
                    quit(1)
            else:
                echo "Invalid option: ", key, "\n"
                echo HelpMessage
                quit(1)
    if result.inputFiles.len() == 0:
        echo "No input files specified.\n"
        echo HelpMessage
        quit(1)


when isMainModule:
    var args = parseCommandLineArguments()
    for file in args.inputFiles:
        let
            content = readFile(file)
            formattedContent = formatContent(content)
        if args.inPlace:
            writeFile(file, formattedContent)
        else:
            echo formattedContent
